registry fail to start (4.2-2019-08-08-070705)

Registry fails to start after nightly build on azure

```
openshift-image-registry                                cluster-image-registry-operator-c8d6678cd-v8zmp                   0/1     CrashLoopBackOff   10         33m
```

```
[mjudeiki@redhat installer]$ oc describe pod cluster-image-registry-operator-c8d6678cd-v8zmp 
Name:               cluster-image-registry-operator-c8d6678cd-v8zmp
Namespace:          openshift-image-registry
Priority:           2000000000
PriorityClassName:  system-cluster-critical
Node:               mjtest-2gslw-master-0/10.0.0.7
Start Time:         Thu, 08 Aug 2019 09:40:26 +0100
Labels:             name=cluster-image-registry-operator
                    pod-template-hash=c8d6678cd
Annotations:        openshift.io/scc: restricted
Status:             Running
IP:                 10.129.0.20
Controlled By:      ReplicaSet/cluster-image-registry-operator-c8d6678cd
Containers:
  cluster-image-registry-operator:
    Container ID:  cri-o://c07f5c2d6d41e4ad5e78b34f50c0081202d065e2d5ce3a5af1029ac3c518c632
    Image:         registry.svc.ci.openshift.org/origin/4.2-2019-08-08-070705@sha256:1f03caa36cec488f8b8a854c81577f7e89e822623396e83cbb9275c8090643db
    Image ID:      registry.svc.ci.openshift.org/origin/4.2-2019-08-08-070705@sha256:1f03caa36cec488f8b8a854c81577f7e89e822623396e83cbb9275c8090643db
    Port:          60000/TCP
    Host Port:     0/TCP
    Command:
      cluster-image-registry-operator
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    2
      Started:      Thu, 08 Aug 2019 10:11:11 +0100
      Finished:     Thu, 08 Aug 2019 10:11:30 +0100
    Ready:          False
    Restart Count:  10
    Requests:
      cpu:  10m
    Environment:
      RELEASE_VERSION:  4.2.0-0.okd-2019-08-08-070705
      WATCH_NAMESPACE:  openshift-image-registry (v1:metadata.namespace)
      OPERATOR_NAME:    cluster-image-registry-operator
      IMAGE:            registry.svc.ci.openshift.org/origin/4.2-2019-08-08-070705@sha256:e252d48cc5108337c1a1b4a90366898da3454b50e4407a620d8e7ba146758e15
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from cluster-image-registry-operator-token-dc5kt (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  cluster-image-registry-operator-token-dc5kt:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  cluster-image-registry-operator-token-dc5kt
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  node-role.kubernetes.io/master=
Tolerations:     node-role.kubernetes.io/master:NoSchedule
                 node.kubernetes.io/memory-pressure:NoSchedule
                 node.kubernetes.io/not-ready:NoExecute for 120s
                 node.kubernetes.io/unreachable:NoExecute for 120s
Events:
  Type     Reason     Age                    From                            Message
  ----     ------     ----                   ----                            -------
  Normal   Scheduled  34m                    default-scheduler               Successfully assigned openshift-image-registry/cluster-image-registry-operator-c8d6678cd-v8zmp to mjtest-2gslw-master-0
  Normal   Pulling    34m                    kubelet, mjtest-2gslw-master-0  Pulling image "registry.svc.ci.openshift.org/origin/4.2-2019-08-08-070705@sha256:1f03caa36cec488f8b8a854c81577f7e89e822623396e83cbb9275c8090643db"
  Normal   Pulled     33m                    kubelet, mjtest-2gslw-master-0  Successfully pulled image "registry.svc.ci.openshift.org/origin/4.2-2019-08-08-070705@sha256:1f03caa36cec488f8b8a854c81577f7e89e822623396e83cbb9275c8090643db"
  Normal   Created    30m (x5 over 33m)      kubelet, mjtest-2gslw-master-0  Created container cluster-image-registry-operator
  Normal   Started    30m (x5 over 33m)      kubelet, mjtest-2gslw-master-0  Started container cluster-image-registry-operator
  Normal   Pulled     28m (x5 over 33m)      kubelet, mjtest-2gslw-master-0  Container image "registry.svc.ci.openshift.org/origin/4.2-2019-08-08-070705@sha256:1f03caa36cec488f8b8a854c81577f7e89e822623396e83cbb9275c8090643db" already present on machine
  Warning  BackOff    3m54s (x110 over 33m)  kubelet, mjtest-2gslw-master-0  Back-off restarting failed container
```
Logs: 
```
[mjudeiki@redhat installer]$ oc logs -f cluster-image-registry-operator-c8d6678cd-v8zmp 
I0808 09:11:11.318896       1 main.go:20] Cluster Image Registry Operator Version: 4.0.0-300-g4a57a26-dirty
I0808 09:11:11.318989       1 main.go:21] Go Version: go1.12.5
I0808 09:11:11.318996       1 main.go:22] Go OS/Arch: linux/amd64
I0808 09:11:11.322967       1 controller.go:473] waiting for informer caches to sync
I0808 09:11:12.925302       1 controller.go:482] started events processor
I0808 09:11:13.381040       1 azure.go:144] attempt to create azure storage account mjtest2gslwmtnn4 (resourceGroup="mjtest-2gslw-rg", location="eastus")...
I0808 09:11:30.889054       1 azure.go:174] azure storage account mjtest2gslwmtnn4 has been created
E0808 09:11:30.925871       1 runtime.go:69] Observed a panic: "slice bounds out of range" (runtime error: slice bounds out of range)
/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/panic.go:522
/usr/local/go/src/runtime/panic.go:54
/go/src/github.com/openshift/cluster-image-registry-operator/pkg/storage/azure/azure.go:471
/go/src/github.com/openshift/cluster-image-registry-operator/pkg/resource/generator.go:102
/go/src/github.com/openshift/cluster-image-registry-operator/pkg/resource/generator.go:143
/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:121
/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:159
/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:249
/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:256
/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152
/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153
/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:1337
panic: runtime error: slice bounds out of range [recovered]
	panic: runtime error: slice bounds out of range

goroutine 253 [running]:
github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x105
panic(0x1863020, 0x3052d20)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/openshift/cluster-image-registry-operator/pkg/storage/azure.(*driver).CreateStorage(0xc000a01ec0, 0xc0005d0840, 0xc0005c2101, 0x1e66f20)
	/go/src/github.com/openshift/cluster-image-registry-operator/pkg/storage/azure/azure.go:471 +0x18da
github.com/openshift/cluster-image-registry-operator/pkg/resource.(*Generator).syncStorage(0xc0000be280, 0xc0005d0840, 0x41365a, 0xc000a41580)
	/go/src/github.com/openshift/cluster-image-registry-operator/pkg/resource/generator.go:102 +0xc9
github.com/openshift/cluster-image-registry-operator/pkg/resource.(*Generator).Apply(0xc0000be280, 0xc0005d0840, 0x0, 0x0)
	/go/src/github.com/openshift/cluster-image-registry-operator/pkg/resource/generator.go:143 +0x4d
github.com/openshift/cluster-image-registry-operator/pkg/operator.(*Controller).createOrUpdateResources(0xc0005c4180, 0xc0005d0840, 0x7, 0xc0000f6801)
	/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:121 +0x1a9
github.com/openshift/cluster-image-registry-operator/pkg/operator.(*Controller).sync(0xc0005c4180, 0x1e765b0, 0xc0000be320)
	/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:159 +0x108e
github.com/openshift/cluster-image-registry-operator/pkg/operator.(*Controller).eventProcessor.func1(0xc0005c4180, 0x17bbd00, 0x1dde3d0)
	/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:249 +0x98
github.com/openshift/cluster-image-registry-operator/pkg/operator.(*Controller).eventProcessor(0xc0005c4180)
	/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:256 +0x45
github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000a41560)
	/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x54
github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000a41560, 0x3b9aca00, 0x0, 0x1, 0xc0000886c0)
	/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc000a41560, 0x3b9aca00, 0xc0000886c0)
	/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/openshift/cluster-image-registry-operator/pkg/operator.(*Controller).Run
	/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:480 +0xf5e
[mjudeiki@redhat installer]$ 
```

/kind bug

--- a/pkg/storage/azure/azure.go
+++ b/pkg/storage/azure/azure.go
@@ -18,7 +18,6 @@ import (
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/util/rand"
-	"k8s.io/apimachinery/pkg/util/uuid"
 	"k8s.io/client-go/rest"
 	"k8s.io/klog"
 
@@ -463,12 +462,14 @@ func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
 				util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapiv1.ConditionFalse, storageExistsReasonAzureError, fmt.Sprintf("Unable to get account primary key: %s", err))
 				return err
 			}
-
-			for i := 0; i < 5000; i++ {
+			const numRetries = 5000
+			for i := 0; i < numRetries; i++ {
 				// If the bucket name is blank, let's generate one
 				if len(d.Config.Container) == 0 {
 					// Container name must be between 3 and 63 characters long
-					d.Config.Container = fmt.Sprintf("%s-%s-%s", infra.Status.InfrastructureName, imageregistryv1.ImageRegistryName, strings.Replace(string(uuid.NewUUID()), "-", "", -1))[0:62]
+					if d.Config.Container, err = util.GenerateStorageName(d.Listers, ""); err != nil {
+						return err
+					}
 				}
 
 				err = d.createStorageContainer(d.Config.AccountName, key, d.Config.Container)

--- a/pkg/storage/gcs/gcs.go
+++ b/pkg/storage/gcs/gcs.go
@@ -5,11 +5,9 @@ import (
 	"fmt"
 	"reflect"
 	"strconv"
-	"strings"
 
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
-	"k8s.io/apimachinery/pkg/util/uuid"
 	"k8s.io/client-go/rest"
 
 	gstorage "cloud.google.com/go/storage"
@@ -204,11 +202,6 @@ func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
 		return err
 	}
 
-	infra, err := util.GetInfrastructure(d.Listers)
-	if err != nil {
-		return err
-	}
-
 	// If a bucket name is supplied, and it already exists and we can access it
 	// just update the config
 	var bucket *gstorage.BucketHandle
@@ -239,10 +232,13 @@ func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
 		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapi.ConditionTrue, "GCS Bucket Exists", "User supplied GCS bucket exists and is accessible")
 
 	} else {
-		for i := 0; i < 5000; i++ {
+		const numRetries = 5000
+		for i := 0; i < numRetries; i++ {
 			// If the bucket name is blank, let's generate one
 			if len(d.Config.Bucket) == 0 {
-				d.Config.Bucket = fmt.Sprintf("%s-%s-%s-%s", infra.Status.InfrastructureName, imageregistryv1.ImageRegistryName, d.Config.Region, strings.Replace(string(uuid.NewUUID()), "-", "", -1))[0:62]
+				if d.Config.Bucket, err = util.GenerateStorageName(d.Listers, d.Config.Region); err != nil {
+					return err
+				}
 			}
 			bucketAttrs := gstorage.BucketAttrs{Location: d.Config.Region}
 			bucket = gclient.Bucket(d.Config.Bucket)

--- a/pkg/storage/s3/s3.go
+++ b/pkg/storage/s3/s3.go
@@ -6,7 +6,6 @@ import (
 	"net/http"
 	"net/url"
 	"reflect"
-	"strings"
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/awserr"
@@ -19,7 +18,6 @@ import (
 
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
-	"k8s.io/apimachinery/pkg/util/uuid"
 	"k8s.io/client-go/rest"
 
 	configapiv1 "github.com/openshift/api/config/v1"
@@ -358,10 +356,13 @@ func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
 	} else {
 		generatedName := false
 		// Retry up to 5000 times if we get a naming conflict
-		for i := 0; i < 5000; i++ {
+		const numRetries = 5000
+		for i := 0; i < numRetries; i++ {
 			// If the bucket name is blank, let's generate one
 			if len(d.Config.Bucket) == 0 {
-				d.Config.Bucket = fmt.Sprintf("%s-%s-%s-%s", infra.Status.InfrastructureName, imageregistryv1.ImageRegistryName, d.Config.Region, strings.Replace(string(uuid.NewUUID()), "-", "", -1))[0:62]
+				if d.Config.Bucket, err = util.GenerateStorageName(d.Listers, d.Config.Region); err != nil {
+					return err
+				}
 				generatedName = true
 			}
 
@@ -441,6 +442,7 @@ func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
 		_, err := svc.PutBucketTaggingWithContext(d.Context, &s3.PutBucketTaggingInput{
 			Bucket: aws.String(d.Config.Bucket),
 			Tagging: &s3.Tagging{
+
 				TagSet: []*s3.Tag{
 					{
 						Key:   aws.String("kubernetes.io/cluster/" + infra.Status.InfrastructureName),

--- a/pkg/storage/swift/swift.go
+++ b/pkg/storage/swift/swift.go
@@ -367,7 +367,7 @@ func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
 		if err != nil {
 			// If the error is not ErrResourceNotFound
 			// return the error
-			if _, ok := err.(*gophercloud.ErrResourceNotFound); !ok {
+			if _, ok := err.(gophercloud.ErrDefault404); !ok {
 				util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapi.ConditionFalse, "Unable to check if container exists", fmt.Sprintf("Error occurred checking if container exists: %v", err))
 				return err
 			}

--- a/pkg/storage/swift/swift.go
+++ b/pkg/storage/swift/swift.go
@@ -2,7 +2,6 @@ package swift
 
 import (
 	"fmt"
-	"math/rand"
 	"reflect"
 	"strings"
 
@@ -302,6 +301,16 @@ func (d *driver) ensureAuthURLHasAPIVersion() error {
 	return nil
 }
 
+func (d *driver) containerExists(client *gophercloud.ServiceClient, containerName string) error {
+	if len(containerName) == 0 {
+		return nil
+	}
+	_, err := containers.Get(client, containerName, containers.GetOpts{}).Extract()
+
+	return err
+
+}
+
 func (d *driver) StorageExists(cr *imageregistryv1.Config) (bool, error) {
 	client, err := d.getSwiftClient(cr)
 	if err != nil {
@@ -309,9 +318,13 @@ func (d *driver) StorageExists(cr *imageregistryv1.Config) (bool, error) {
 		return false, err
 	}
 
-	_, err = containers.Get(client, cr.Spec.Storage.Swift.Container, containers.GetOpts{}).Extract()
+	err = d.containerExists(client, cr.Spec.Storage.Swift.Container)
 	if err != nil {
-		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapi.ConditionFalse, "Storage does not exist", err.Error())
+		if serr, ok := err.(*gophercloud.ErrResourceNotFound); ok {
+			util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapi.ConditionFalse, "Storage does not exist", serr.Error())
+			return false, nil
+		}
+		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapi.ConditionUnknown, "Unknown error occurred", err.Error())
 		return false, err
 	}
 
@@ -328,14 +341,6 @@ func (d *driver) StorageChanged(cr *imageregistryv1.Config) bool {
 	return false
 }
 
-func generateContainerName(clusterID string) string {
-	bytes := make([]byte, 8)
-	for i := 0; i < 8; i++ {
-		bytes[i] = byte(65 + rand.Intn(25)) // A=65 and Z=65+25
-	}
-	return clusterID + "-image-registry-" + string(bytes)
-}
-
 func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
 	client, err := d.getSwiftClient(cr)
 	if err != nil {
@@ -348,33 +353,62 @@ func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
 		return fmt.Errorf("failed to get cluster infrastructure info: %v", err)
 	}
 
-	// Generate new container name if it wasn't provided.
-	// The name has cluster ID as a prefix plus "image-registry" suffix, which is complemented
-	// by 8 capital latin letters
-	// Example of a generated name: user-298xw-image-registry-FHEIBGDD
-	if cr.Spec.Storage.Swift.Container == "" {
-		cr.Spec.Storage.Swift.Container = generateContainerName(infra.Status.InfrastructureName)
+	generatedName := false
+	const numRetries = 5000
+	for i := 0; i < numRetries; i++ {
+		if len(cr.Spec.Storage.Swift.Container) == 0 {
+			if cr.Spec.Storage.Swift.Container, err = util.GenerateStorageName(d.Listers, ""); err != nil {
+				return err
+			}
+			generatedName = true
+		}
+
+		err = d.containerExists(client, cr.Spec.Storage.Swift.Container)
+		if err != nil {
+			// If the error is not ErrResourceNotFound
+			// return the error
+			if _, ok := err.(*gophercloud.ErrResourceNotFound); !ok {
+				util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapi.ConditionFalse, "Unable to check if container exists", fmt.Sprintf("Error occurred checking if container exists: %v", err))
+				return err
+			}
+			// If the error is ErrResourceNotFound
+			// fall through to the container creation
+		}
+		// If we were supplied a container name and it exists
+		// we can skip the create
+		if !generatedName && err == nil {
+			util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapi.ConditionTrue, "Container exists", "User supplied container already exists")
+			break
+		}
+		// If we generated a container name and it exists
+		// let's try again
+		if generatedName && err == nil {
+			cr.Spec.Storage.Swift.Container = ""
+			continue
+		}
+
+		createOps := containers.CreateOpts{
+			Metadata: map[string]string{
+				"Openshiftclusterid": infra.Status.InfrastructureName,
+				"Name":               cr.Spec.Storage.Swift.Container,
+			},
+		}
+
+		_, err = containers.Create(client, cr.Spec.Storage.Swift.Container, createOps).Extract()
+		if err != nil {
+			util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapi.ConditionFalse, "Creation Failed", err.Error())
+			cr.Status.StorageManaged = false
+			return err
+		}
+
+		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapi.ConditionTrue, "Swift Container Created", "")
+
+		cr.Status.StorageManaged = true
+		cr.Status.Storage.Swift = d.Config.DeepCopy()
+		cr.Spec.Storage.Swift = d.Config.DeepCopy()
+
+		break
 	}
-
-	createOps := containers.CreateOpts{
-		Metadata: map[string]string{
-			"Openshiftclusterid": infra.Status.InfrastructureName,
-			"Name":               cr.Spec.Storage.Swift.Container,
-		},
-	}
-
-	_, err = containers.Create(client, cr.Spec.Storage.Swift.Container, createOps).Extract()
-	if err != nil {
-		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapi.ConditionFalse, "Creation Failed", err.Error())
-		cr.Status.StorageManaged = false
-		return err
-	}
-
-	util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapi.ConditionTrue, "Swift Container Created", "")
-
-	cr.Status.StorageManaged = true
-	cr.Status.Storage.Swift = d.Config.DeepCopy()
-	cr.Spec.Storage.Swift = d.Config.DeepCopy()
 
 	return nil
 }

--- a/pkg/storage/swift/swift_test.go
+++ b/pkg/storage/swift/swift_test.go
@@ -219,20 +219,20 @@ func TestSwiftCreateStorageNativeSecret(t *testing.T) {
 		// to check if container with name exists
 		if numRequests == 0 {
 			th.TestMethod(t, r, "HEAD")
+			th.TestHeader(t, r, "Accept", "application/json")
+			w.WriteHeader(http.StatusNotFound)
 			numRequests++
 		} else {
 			// Second request should be the actual create
 			th.TestMethod(t, r, "PUT")
+			th.TestHeader(t, r, "Accept", "application/json")
+
+			w.Header().Set("Content-Length", "0")
+			w.Header().Set("Content-Type", "text/html; charset=UTF-8")
+			w.Header().Set("Date", "Wed, 17 Aug 2016 19:25:43 GMT")
+			w.Header().Set("X-Trans-Id", "tx554ed59667a64c61866f1-0058b4ba37")
+			w.WriteHeader(http.StatusNoContent)
 		}
-
-		th.TestHeader(t, r, "Accept", "application/json")
-
-		w.Header().Set("Content-Length", "0")
-		w.Header().Set("Content-Type", "text/html; charset=UTF-8")
-		w.Header().Set("Date", "Wed, 17 Aug 2016 19:25:43 GMT")
-		w.Header().Set("X-Trans-Id", "tx554ed59667a64c61866f1-0058b4ba37")
-		w.WriteHeader(http.StatusNoContent)
-
 	})
 
 	d, installConfig := mockConfig(false, th.Endpoint()+"v3", MockUPISecretNamespaceLister{})
@@ -370,18 +370,19 @@ func TestSwiftCreateStorageCloudConfig(t *testing.T) {
 	th.Mux.HandleFunc("/"+container, func(w http.ResponseWriter, r *http.Request) {
 		if numRequests == 0 {
 			th.TestMethod(t, r, "HEAD")
+			th.TestHeader(t, r, "Accept", "application/json")
+			w.WriteHeader(http.StatusNotFound)
 			numRequests++
 		} else {
 			th.TestMethod(t, r, "PUT")
+			th.TestHeader(t, r, "Accept", "application/json")
+
+			w.Header().Set("Content-Length", "0")
+			w.Header().Set("Content-Type", "text/html; charset=UTF-8")
+			w.Header().Set("Date", "Wed, 17 Aug 2016 19:25:43 GMT")
+			w.Header().Set("X-Trans-Id", "tx554ed59667a64c61866f1-0058b4ba37")
+			w.WriteHeader(http.StatusNoContent)
 		}
-
-		th.TestHeader(t, r, "Accept", "application/json")
-
-		w.Header().Set("Content-Length", "0")
-		w.Header().Set("Content-Type", "text/html; charset=UTF-8")
-		w.Header().Set("Date", "Wed, 17 Aug 2016 19:25:43 GMT")
-		w.Header().Set("X-Trans-Id", "tx554ed59667a64c61866f1-0058b4ba37")
-		w.WriteHeader(http.StatusNoContent)
 	})
 
 	d, installConfig := mockConfig(false, th.Endpoint()+"v3", MockIPISecretNamespaceLister{})
@@ -581,30 +582,26 @@ func TestSwiftEndpointTypeObjectStore(t *testing.T) {
 	defer th.TeardownHTTP()
 	handleAuthentication(t, "object-store")
 
-	numRequests := 0
-
 	th.Mux.HandleFunc("/"+container, func(w http.ResponseWriter, r *http.Request) {
-		if numRequests == 0 {
-			th.TestMethod(t, r, "HEAD")
-			numRequests++
-		} else {
-			th.TestMethod(t, r, "PUT")
-		}
+		th.TestMethod(t, r, "HEAD")
 		th.TestHeader(t, r, "Accept", "application/json")
-
-		w.Header().Set("Content-Length", "0")
-		w.Header().Set("Content-Type", "text/html; charset=UTF-8")
+		w.Header().Set("Accept-Ranges", "bytes")
+		w.Header().Set("Content-Type", "application/json; charset=utf-8")
 		w.Header().Set("Date", "Wed, 17 Aug 2016 19:25:43 GMT")
-		w.Header().Set("X-Trans-Id", "tx554ed59667a64c61866f1-0058b4ba37")
+		w.Header().Set("X-Container-Bytes-Used", "100")
+		w.Header().Set("X-Container-Object-Count", "4")
+		w.Header().Set("X-Container-Read", "test")
+		w.Header().Set("X-Container-Write", "test2,user4")
+		w.Header().Set("X-Timestamp", "1471298837.95721")
+		w.Header().Set("X-Trans-Id", "tx554ed59667a64c61866f1-0057b4ba37")
+		w.Header().Set("X-Storage-Policy", "test_policy")
 		w.WriteHeader(http.StatusNoContent)
 	})
 
 	d, installConfig := mockConfig(false, th.Endpoint()+"v3", MockUPISecretNamespaceLister{})
 
-	d.CreateStorage(&installConfig)
+	res, err := d.StorageExists(&installConfig)
 
-	th.AssertEquals(t, true, installConfig.Status.StorageManaged)
-	th.AssertEquals(t, "StorageExists", installConfig.Status.Conditions[0].Type)
-	th.AssertEquals(t, operatorapi.ConditionTrue, installConfig.Status.Conditions[0].Status)
-	th.AssertEquals(t, container, installConfig.Status.Storage.Swift.Container)
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, true, res)
 }

--- a/pkg/storage/swift/swift_test.go
+++ b/pkg/storage/swift/swift_test.go
@@ -212,8 +212,19 @@ func TestSwiftCreateStorageNativeSecret(t *testing.T) {
 	defer th.TeardownHTTP()
 	handleAuthentication(t, "container")
 
+	numRequests := 0
+
 	th.Mux.HandleFunc("/"+container, func(w http.ResponseWriter, r *http.Request) {
-		th.TestMethod(t, r, "PUT")
+		// The first request should be a head request
+		// to check if container with name exists
+		if numRequests == 0 {
+			th.TestMethod(t, r, "HEAD")
+			numRequests++
+		} else {
+			// Second request should be the actual create
+			th.TestMethod(t, r, "PUT")
+		}
+
 		th.TestHeader(t, r, "Accept", "application/json")
 
 		w.Header().Set("Content-Length", "0")
@@ -221,6 +232,7 @@ func TestSwiftCreateStorageNativeSecret(t *testing.T) {
 		w.Header().Set("Date", "Wed, 17 Aug 2016 19:25:43 GMT")
 		w.Header().Set("X-Trans-Id", "tx554ed59667a64c61866f1-0058b4ba37")
 		w.WriteHeader(http.StatusNoContent)
+
 	})
 
 	d, installConfig := mockConfig(false, th.Endpoint()+"v3", MockUPISecretNamespaceLister{})
@@ -353,8 +365,16 @@ func TestSwiftCreateStorageCloudConfig(t *testing.T) {
 		cloudSecretKey: fakeCloudsYAMLData,
 	}
 
+	numRequests := 0
+
 	th.Mux.HandleFunc("/"+container, func(w http.ResponseWriter, r *http.Request) {
-		th.TestMethod(t, r, "PUT")
+		if numRequests == 0 {
+			th.TestMethod(t, r, "HEAD")
+			numRequests++
+		} else {
+			th.TestMethod(t, r, "PUT")
+		}
+
 		th.TestHeader(t, r, "Accept", "application/json")
 
 		w.Header().Set("Content-Length", "0")
@@ -561,8 +581,15 @@ func TestSwiftEndpointTypeObjectStore(t *testing.T) {
 	defer th.TeardownHTTP()
 	handleAuthentication(t, "object-store")
 
+	numRequests := 0
+
 	th.Mux.HandleFunc("/"+container, func(w http.ResponseWriter, r *http.Request) {
-		th.TestMethod(t, r, "PUT")
+		if numRequests == 0 {
+			th.TestMethod(t, r, "HEAD")
+			numRequests++
+		} else {
+			th.TestMethod(t, r, "PUT")
+		}
 		th.TestHeader(t, r, "Accept", "application/json")
 
 		w.Header().Set("Content-Length", "0")
@@ -580,8 +607,4 @@ func TestSwiftEndpointTypeObjectStore(t *testing.T) {
 	th.AssertEquals(t, "StorageExists", installConfig.Status.Conditions[0].Type)
 	th.AssertEquals(t, operatorapi.ConditionTrue, installConfig.Status.Conditions[0].Status)
 	th.AssertEquals(t, container, installConfig.Status.Storage.Swift.Container)
-}
-
-func TestSwiftGenerateContainerName(t *testing.T) {
-	th.AssertEquals(t, false, generateContainerName("image_registry_") == generateContainerName("image_registry_"))
 }

--- a/pkg/storage/util/util.go
+++ b/pkg/storage/util/util.go
@@ -2,6 +2,7 @@ package util
 
 import (
 	"fmt"
+	"math/rand"
 	"strings"
 
 	corev1 "k8s.io/api/core/v1"
@@ -103,4 +104,53 @@ func GetValueFromSecret(sec *corev1.Secret, key string) (string, error) {
 		return string(v), nil
 	}
 	return "", fmt.Errorf("secret %q does not contain required key %q", fmt.Sprintf("%s/%s", sec.Namespace, sec.Name), key)
+}
+
+// GenerateStorageName generates a unique name for the storage
+// medium that the registry will use
+func GenerateStorageName(listers *regopclient.Listers, additionalInfo ...string) (string, error) {
+
+	// Get the infrastructure name
+	infra, err := GetInfrastructure(listers)
+	if err != nil {
+		return "", err
+	}
+
+	// A slice to store the parts of our name
+	var parts []string
+
+	// Put the infrastructure name first
+	parts = append(parts, infra.Status.InfrastructureName)
+
+	// Image Registry Name second
+	parts = append(parts, imageregistryv1.ImageRegistryName)
+
+	// Additional information provided to the function third
+	for _, i := range additionalInfo {
+		if len(i) != 0 {
+			parts = append(parts, i)
+		}
+	}
+
+	// Join the slice together with dashes
+	name := strings.Join(parts, "-")
+
+	// Check the length and pad or truncate as needed
+	switch {
+	case len(name) < 62:
+		padding := 62 - len(name) - 1
+		bytes := make([]byte, padding)
+		for i := 0; i < padding; i++ {
+			bytes[i] = byte(97 + rand.Intn(25)) // a=97 and z=97+25
+		}
+		name = fmt.Sprintf("%s-%s", name, string(bytes))
+	case len(name) > 62:
+		name = name[0:62]
+		if strings.HasSuffix(name, "-") {
+			name = name[0:61] + string(byte(97+rand.Intn(25)))
+		}
+	}
+
+	return name, nil
+
 }

--- a/pkg/storage/util/util_test.go
+++ b/pkg/storage/util/util_test.go
@@ -1,0 +1,86 @@
+package util
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	configv1 "github.com/openshift/api/config/v1"
+	imageregistryv1 "github.com/openshift/cluster-image-registry-operator/pkg/apis/imageregistry/v1"
+	regopclient "github.com/openshift/cluster-image-registry-operator/pkg/client"
+)
+
+var (
+	infrastructureName = "jsmith-xhv4"
+	prefix             = fmt.Sprintf("%s-%s", infrastructureName, imageregistryv1.ImageRegistryName)
+)
+
+type MockInfrastructureLister struct {
+}
+
+func (m MockInfrastructureLister) Get(name string) (*configv1.Infrastructure, error) {
+	return &configv1.Infrastructure{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "cluster",
+		},
+		Status: configv1.InfrastructureStatus{
+			InfrastructureName: infrastructureName,
+			PlatformStatus:     &configv1.PlatformStatus{},
+		},
+	}, nil
+}
+
+func (m MockInfrastructureLister) List(selector labels.Selector) ([]*configv1.Infrastructure, error) {
+	var list []*configv1.Infrastructure
+	return list, nil
+}
+
+func TestGenerateStorageName(t *testing.T) {
+	l := regopclient.Listers{Infrastructures: MockInfrastructureLister{}}
+	tests := [][]string{
+		// test with nil
+		nil,
+		// test with no additionals
+		{},
+		// test with two empty strings
+		{"", ""},
+		// test one additional
+		{"test1"},
+		// test two additionals
+		{"test1", "test2"},
+		// test really long additionals
+		{"abcdefghijklmnopqrstuvwxyz", "abcdefghijklmnopqrstuvwxyz"},
+	}
+
+	multiDash := regexp.MustCompile(`-{2,}`)
+	replaceMultiDash := regexp.MustCompile(`-{1,}`)
+	for _, test := range tests {
+		n, err := GenerateStorageName(&l, test...)
+		if err != nil {
+			t.Errorf("%v", err)
+		}
+		wantedPrefix := replaceMultiDash.ReplaceAllString(fmt.Sprintf("%s-%s", prefix, strings.Join(test, "-")), "-")
+		if len(wantedPrefix) > 62 {
+			wantedPrefix = wantedPrefix[0:62]
+		}
+
+		// Name should have the wanted prefix
+		if !strings.HasPrefix(n, wantedPrefix) {
+			t.Errorf("name should have the prefix %s, but was %s instead", wantedPrefix, n)
+		}
+
+		// Name should be exactly 62 characters long
+		if len(n) != 62 {
+			t.Errorf("name should be exactly 62 characters long, but was %d instead: %s", len(n), n)
+		}
+
+		// Name should not have multiple dashes together
+		if multiDash.MatchString(n) {
+			t.Errorf("name should not include a double dash: %s", n)
+		}
+	}
+}

